tmt/frameworks/shell.py

```diff
@@ -1,12 +1,26 @@
+import re
 from typing import Optional
 
 import tmt.log
 import tmt.result
 import tmt.steps.execute
 import tmt.utils
-from tmt.frameworks import TestFramework, provides_framework
+from tmt.frameworks import TestFramework, provides_framework, save_test_failures
 from tmt.result import ResultOutcome
 from tmt.steps.execute import TEST_OUTPUT_FILENAME, TestInvocation
+from tmt.utils import Path
+
+
+def _extract_failure(log_path: Path, invocation: 'TestInvocation') -> Optional[str]:
+    if not log_path.is_file():
+        return None
+    try:
+        log = invocation.phase.read(log_path)
+    except tmt.utils.FileError:
+        return None
+
+    failures = re.findall(r'.*\b(?:error|fail)\b.*', log, re.IGNORECASE | re.MULTILINE)
+    return '\n'.join(failures) if failures else None
```

Comment:
Same here, return a list. Here I think we can return individual findings, i.e. not bundling them into a single string and single item of the final list.